pkg/networking/middleware/retry_middleware.go

```diff
@@ -0,0 +1,108 @@
+package middleware
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff/v5"
+
+	"github.com/snyk/go-application-framework/pkg/configuration"
+)
+
+const defaultRetryCount uint = 3
+const configurationKeyRetryCount = "INTERNAL_MAX_NET_RETRIES"
+const defaultRetryAfterSeconds = 5
+
+type RetryMiddleware struct {
+	next   http.RoundTripper
+	config configuration.Configuration
+}
+
+func NewRetryMiddleware(roundTriper http.RoundTripper, config configuration.Configuration) *RetryMiddleware {
+	return &RetryMiddleware{
+		next:   roundTriper,
+		config: config,
+	}
+}
+
+func (rm RetryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	var finalResponse *http.Response = nil
+	var finalError error = nil
+	var localBodyBuffer []byte
+	var maxRetryCount = defaultRetryCount
+
+	// if a body is available, create a local copy to be able to use it multiple times
+	if req.Body != nil {
```

Comment:
suggestion: as this is a library and will be heavily used, I would suggest to have limits on body size. Imagine having a POST request with a 2 GB repo source code thingy (by accident or not). I see two options: either limit retries to a max request size or swap to filesystem after a threshold.